core/crypto/kem/adapter/kem.go

```diff
@@ -0,0 +1,249 @@
+package adapter
+
+import (
+	"crypto/hmac"
+
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/cloudflare/circl/kem"
+
+	"github.com/katzenpost/katzenpost/core/crypto/nike"
+	"github.com/katzenpost/katzenpost/core/crypto/rand"
+)
+
+// PublicKey is an adapter for nike.PublicKey to kem.PublicKey.
+type PublicKey struct {
+	publicKey nike.PublicKey
+	scheme    *Scheme
+}
+
+func (p *PublicKey) Scheme() kem.Scheme {
+	return p.scheme
+}
+
+func (p *PublicKey) MarshalBinary() ([]byte, error) {
+	return p.publicKey.MarshalBinary()
+}
+
+func (p *PublicKey) Equal(pubkey kem.PublicKey) bool {
+	if pubkey.(*PublicKey).scheme != p.scheme {
+		return false
+	}
+	return hmac.Equal(pubkey.(*PublicKey).publicKey.Bytes(), p.publicKey.Bytes())
+}
+
+// PrivateKey is an adapter for nike.PrivateKey to kem.PrivateKey.
+type PrivateKey struct {
+	privateKey nike.PrivateKey
+	scheme     *Scheme
+}
+
+func (p *PrivateKey) Scheme() kem.Scheme {
+	return p.scheme
+}
+
+func (p *PrivateKey) MarshalBinary() ([]byte, error) {
+	return p.privateKey.MarshalBinary()
+}
+
+func (p *PrivateKey) Equal(privkey kem.PrivateKey) bool {
+	if privkey.(*PrivateKey).scheme != p.scheme {
+		return false
+	}
+	return hmac.Equal(privkey.(*PrivateKey).privateKey.Bytes(), p.privateKey.Bytes())
+}
+
+func (p *PrivateKey) Public() kem.PublicKey {
+	return &PublicKey{
+		publicKey: p.privateKey.Public(),
+		scheme:    p.scheme,
+	}
+}
+
+// Scheme is an adapter for nike.Scheme to kem.Scheme.
+type Scheme struct {
+	nike nike.Scheme
+}
+
+var _ kem.Scheme = (*Scheme)(nil)
+var _ kem.PublicKey = (*PublicKey)(nil)
+var _ kem.PrivateKey = (*PrivateKey)(nil)
+
+// FromNIKE creates a new KEM adapter Scheme
+// using the given NIKE Scheme.
+func FromNIKE(nike nike.Scheme) *Scheme {
+	return &Scheme{
+		nike: nike,
+	}
+}
+
+// Name of the scheme
+func (a *Scheme) Name() string {
+	return a.nike.Name()
+}
+
+// GenerateKeyPair creates a new key pair.
+func (a *Scheme) GenerateKeyPair() (kem.PublicKey, kem.PrivateKey, error) {
+	pubkey, privkey, err := a.nike.GenerateKeyPair()
```

Comment:
Do we want deterministic functions for these like we have `GenerateKeyPairFromEntropy` ? Seems like it could be a useful building block for other interesting constructions, but especially for testing.

Reply:
we now have `DeriveKeyPair` in this file